README.md
# fast-geodist

[![PyPI](https://img.shields.io/pypi/v/fast-geodist?style=for-the-badge)](https://pypi.org/project/fast-geodist/)
[![CI](https://img.shields.io/github/workflow/status/mblackgeo/fast-geodist/ci?label=CI&style=for-the-badge)](https://github.com/mblackgeo/fast-geodist/actions)
[![Release](https://img.shields.io/github/workflow/status/mblackgeo/fast-geodist/release?label=RELEASE&style=for-the-badge)](https://github.com/mblackgeo/fast-geodist/actions)
[![Code Style](https://img.shields.io/static/v1?label=code%20style&message=black&color=blue&style=for-the-badge)](https://github.com/psf/black)

An implementation of the [Haversine formula](https://en.wikipedia.org/wiki/Haversine_formula) for calculating [Great Circle distance](https://en.wikipedia.org/wiki/Great-circle_distance) using Rust and [PyO3](https://github.com/PyO3/PyO3).

This package was started as an experiment for learning how to implement functions in Rust and wrap them for use in python. The Haversine Formula is implemented following [georust](https://github.com/georust/geo/blob/main/geo/src/algorithm/haversine_distance.rs), with an array-wise implementation written using [`ndarray::parallel`](https://docs.rs/ndarray/latest/ndarray/parallel/index.html). It is wrapped for python using [Maturin](https://github.com/PyO3/maturin) and integrates [`rust-numpy`](https://github.com/PyO3/rust-numpy) for array operations.

## Installation

```
pip install fast-geodist
```

## Quick start

The package provides two main entry points:

```python
import numpy as np
from fast_geodist import haversine

# Calculate a single distance
# input either a pair of coordinate tuples: (lat1/lon1), (lat2/lon2)
london = (51.51, -0.12)
paris = (48.85, 2.35)
new_york = (40.7, -74.2)

result = haversine(london, paris)

# Calculate multiple distances with lists
result = haversine([(london, paris), (new_york, london)])

# or a numpy array with shape [4, n]
result = haversine(np.array([(*london, *paris), (*new_york, *london)]))
```

## Benchmarks

The results of benchmarking show the rust implementation is **14x faster** than the python implementation. This benchmark computes distances on an array containing 1,000,000 pairs of coordinates (see [`bench.py`](/tests/bench.py)):

```
--------------------------------------------- benchmark: 2 tests ---------------------------------------------
Name (time in ms)              Min                   Max                  Mean                Median
--------------------------------------------------------------------------------------------------------------
test_benchmark_fast       164.0635 (1.0)        171.7663 (1.0)        168.0218 (1.0)        168.4129 (1.0)
test_benchmark_slow     2,335.4281 (14.23)    2,439.7850 (14.20)    2,395.6077 (14.26)    2,406.6356 (14.29)
--------------------------------------------------------------------------------------------------------------
```
Computed on an Intel i7-1165G7.

## Development

Prequisites:

* Python (>=3.7) and make
* [Rust toolchain](https://rustup.rs/)

After cloning the repository, the Makefile includes helpful commands for setting up a development environment, linting, formatting, testing, and benchmarking. Get started as follows:

```shell
# setup a new virtual environment
python -m venv .venv
source .venv/bin/activate

# install the development dependencies
make install

# check other available commands
make help
```

Tooling:

* Cargo and [Pytest](https://docs.pytest.org/en/6.2.x/) are used for testing the Rust and Python code respectively (see [`/tests`](/tests/)).
* Python code is linted using [flake8](https://flake8.pycqa.org/en/latest/) and formatted using [Black](https://github.com/psf/black); rust code with `cargo fmt` and `cargo clippy`.
* [pre-commit](https://pre-commit.com/) is used to run these checks locally before files are pushed to git
* The [Github Actions pipeline](.github/workflows/ci.yml) runs these checks and tests
* [Semantic-release](https://python-semantic-release.readthedocs.io/en/latest/) is used with [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) for automated releasing to PyPI

## References

As well as the excellent PyO3 documentation, the following posts helped with creation of this package:

- https://blog.yossarian.net/2020/08/02/Writing-and-publishing-a-python-module-in-rust
- https://depth-first.com/articles/2020/08/10/python-extensions-in-pure-rust-with-pyo3/
- http://saidvandeklundert.net/learn/2021-11-18-calling-rust-from-python-using-pyo3/
- https://itnext.io/how-to-bind-python-numpy-with-rust-ndarray-2efa5717ed21


CHANGELOG.md
<!--next-version-placeholder-->

## v0.4.0 (2022-09-28)
### Feature
* Performance improvements and rust benchmarks ([#6](https://github.com/mblackgeo/fast-geodist/issues/6)) ([`f482211`](https://github.com/mblackgeo/fast-geodist/commit/f4822115eaacf016c90ef6401da2fe975110d3d0))

## v0.3.0 (2022-09-21)
### Feature
* Rename package to avoid clash on pypi ([`3efe5bf`](https://github.com/mblackgeo/fast-geodist/commit/3efe5bff57296d5e14c3491583ae863ddba68aac))

### Documentation
* Remove completed items from roadmap ([`d4aad2e`](https://github.com/mblackgeo/fast-geodist/commit/d4aad2e14aa0cd717f006d2b5771b1330fc9a477))
* Add badges ([`6dd65c4`](https://github.com/mblackgeo/fast-geodist/commit/6dd65c48d644849752973d182c6af2a17c7ee3f9))
* Rename package ([`dc00675`](https://github.com/mblackgeo/fast-geodist/commit/dc006755489e7a4fd7d3843a081c05d57e35dfc1))
* Add todo for package rename ([`5f5f490`](https://github.com/mblackgeo/fast-geodist/commit/5f5f490ed58383f3265eab933e58569899b49b85))
* Fix spacing in example ([`03d6a8c`](https://github.com/mblackgeo/fast-geodist/commit/03d6a8c0221c8f3e962359a4feb1468ea4ece2dd))
* Add install command ([`dd45b54`](https://github.com/mblackgeo/fast-geodist/commit/dd45b54f055342157f41ea4d71b3a994cfad9184))

## v0.2.0 (2022-09-15)
### Feature
* Rename internal rust package ([`16e61db`](https://github.com/mblackgeo/fast-geodist/commit/16e61db3c3609127470458ef82749d6a6b19344e))

## v0.1.0 (2022-09-15)
### Feature
* First release ([`7a2067d`](https://github.com/mblackgeo/fast-geodist/commit/7a2067dd9688017609463f93df26d2acc388d69a))
* Add CD pipeline and semantic release ([#2](https://github.com/mblackgeo/fast-geodist/issues/2)) ([`f492491`](https://github.com/mblackgeo/fast-geodist/commit/f492491062fe919a6a6b67ae46621917d379993e))
* Add a conveient python wrapper around the rust functions ([`9e12458`](https://github.com/mblackgeo/fast-geodist/commit/9e12458ca4cf495db090e6a571c6142d3b0ef2f6))
* Implement numpy wrapper for array distance calculation ([`0932b84`](https://github.com/mblackgeo/fast-geodist/commit/0932b848ef03a6a0bf9fbe80d1eb58b3a64b5b09))
* Add an array implementation of haversine distance ([`c78bfb1`](https://github.com/mblackgeo/fast-geodist/commit/c78bfb11e805e0b4408610c9fb578989827431ba))
* Increase the amount data used for benchmarking ([`a3fb8a5`](https://github.com/mblackgeo/fast-geodist/commit/a3fb8a552cb654cff8d7142f0403c8b0b11e6fe9))
* Add implementation that works with a vector of coords ([`446ed5f`](https://github.com/mblackgeo/fast-geodist/commit/446ed5f0f605fb6f4603f89d9261d10dc8d494a2))
* Add python haversine function ([`dd74843`](https://github.com/mblackgeo/fast-geodist/commit/dd74843c80e38452d193dc2574d7874ad56006bb))
* Implement algo following georust ([`056b047`](https://github.com/mblackgeo/fast-geodist/commit/056b047c61ab97b4fd59a6aa2cc06e7469a9d933))
* Add wireframe of rust package ([`e6b8e67`](https://github.com/mblackgeo/fast-geodist/commit/e6b8e67d99b6f1c8edcfa817ead4bedb31f4ef0b))

### Fix
* Allocate an empty vector without size ([`c1b0437`](https://github.com/mblackgeo/fast-geodist/commit/c1b04371f3b973af0a6014aca5830f3bd59e7d16))
* Use python 3.7 compatible type hints ([`29f0772`](https://github.com/mblackgeo/fast-geodist/commit/29f07721faec6d3581a4c448273c5151682af638))

### Documentation
* Add benchmark results ([`f37c2d9`](https://github.com/mblackgeo/fast-geodist/commit/f37c2d99cd65921ce3b26e0aacdd3236ce58e819))
* Correct minor typos ([`a6b054f`](https://github.com/mblackgeo/fast-geodist/commit/a6b054fba9fed2181ce620f7de47dcdeda125c04))
* Fix broken link ([`b824b45`](https://github.com/mblackgeo/fast-geodist/commit/b824b45291742de0d183849189af4b898f76c4f4))
* Update notes for use with np.array ([`96201d4`](https://github.com/mblackgeo/fast-geodist/commit/96201d4a264edce9f8633c7d2e19d59a5b908503))
* Remove pre-commit as a prerequisite ([`6828eec`](https://github.com/mblackgeo/fast-geodist/commit/6828eec1acd816e65d0168b0b0ccf5a352358ee6))
* Populate README with full instructions ([`7a240fc`](https://github.com/mblackgeo/fast-geodist/commit/7a240fc5a7851d9ffbf90a37008b37141ed287b3))
* Mark tasks complete ([`dd911e0`](https://github.com/mblackgeo/fast-geodist/commit/dd911e0a3fd7a7a47f3ba549e8345dd0bf6e1432))
* Typo ([`a93ad51`](https://github.com/mblackgeo/fast-geodist/commit/a93ad517db6df30b7a3d993b0f29ba72d181076d))
* Add helpful links ([`141cea4`](https://github.com/mblackgeo/fast-geodist/commit/141cea431cd295703fc73bf91d10155d8e64724e))
* Add more TODOs ([`3725bcc`](https://github.com/mblackgeo/fast-geodist/commit/3725bcc218eb6414756f07bdb55afd031ec4090c))
* Add note to add more docs ([`eeda04a`](https://github.com/mblackgeo/fast-geodist/commit/eeda04a0ae6d33b56408414966d4692d9bba9b47))
* Add TODO list ([`df0e33b`](https://github.com/mblackgeo/fast-geodist/commit/df0e33bd5093ed101aca775cce095e4907d093eb))

### Performance
* Update benchmark ([`86e73d8`](https://github.com/mblackgeo/fast-geodist/commit/86e73d813285bee1d71a28e759658006cba75bc5))
* Update benchmark to work with numpy arrays ([`1ea2c02`](https://github.com/mblackgeo/fast-geodist/commit/1ea2c02cec1da538f76487ec2fef98d07cd1e6d9))
* Preallocate the vector with type and size ([`ab8b2b2`](https://github.com/mblackgeo/fast-geodist/commit/ab8b2b24e9c3fe809fd367b92dda86a94d71b60d))
* Add benchmarks ([`b42a87a`](https://github.com/mblackgeo/fast-geodist/commit/b42a87adfcbeb2e1c039528a602f43d609c7c06c))
* Add pytest-benchmark ([`223c602`](https://github.com/mblackgeo/fast-geodist/commit/223c6026aa4925e8f2c34e1082a8ea7ca814a5b5))

tests/bench.py
import random

import numpy as np

from fast_haversine import haversine_array
from fast_haversine.slow import slow_haversine


def rand_lat() -> float:
    return random.uniform(-90.0, 90.0)


def rand_lng() -> float:
    return random.uniform(-180.0, 180.0)


def create_data() -> np.ndarray:
    random.seed(0)
    return np.array(
        [(rand_lat(), rand_lng(), rand_lat(), rand_lng()) for _ in range(250_000)],
    )


def bench_helper(fast: bool = False):
    data = create_data()
    if fast:
        haversine_array(data)
    else:
        for row in data:
            slow_haversine(*row)


def test_benchmark_fast(benchmark):
    benchmark(bench_helper, fast=True)


def test_benchmark_slow(benchmark):
    benchmark(bench_helper, fast=False)


tests/haversine.rs
use approx::assert_relative_eq;
use fast_haversine::{haversine_distance, haversine_distance_array};
use ndarray::array;

#[test]
fn test_haversine_1() {
    let res = haversine_distance(0.0, 0.0, 1.0, 0.0);
    let expect = 111195.0802335329_f64;
    assert_relative_eq!(res, expect, epsilon = 1.0e-6)
}

#[test]
fn test_haversine_2() {
    let res = haversine_distance(42.3521, -72.1235, 70.612, 72.1260);
    let expect = 7130580.307935911_f64;
    assert_relative_eq!(res, expect, epsilon = 1.0e-6)
}

#[test]
fn test_haversine_array() {
    let arr = array![[42.3521, -72.1235, 70.612, 72.1260], [0.0, 0.0, 1.0, 0.0]];

    let res = haversine_distance_array(&arr.view());
    let expect = [7130580.307935911, 111195.0802335329];

    assert_relative_eq!(res[0], expect[0], epsilon = 1.0e-6);
    assert_relative_eq!(res[1], expect[1], epsilon = 1.0e-6);
}


tests/test_haversine.py
import numpy as np
import pytest

from fast_geodist import haversine
from fast_geodist.slow import slow_haversine


def test_unsupported_input():
    with pytest.raises(NotImplementedError):
        haversine(1.0)


def test_with_tuple():
    res = haversine((42.3521, -72.1235), (70.612, 72.1260))
    assert pytest.approx(res) == 7130580.307935911


def test_with_list_tuples():
    res = haversine(
        [
            ((42.3521, -72.1235), (70.612, 72.1260)),
            ((0.0, 0.0), (1.0, 0.0)),
        ]
    )
    expect = [7130580.307935911, 111195.0802335329]
    assert isinstance(res, list)
    assert pytest.approx(res[0]) == expect[0]
    assert pytest.approx(res[1]) == expect[1]


def test_with_tuple_with_wrong_shape():
    with pytest.raises(NotImplementedError):
        # error with less than 4 elements
        haversine((1, 1))

    with pytest.raises(NotImplementedError):
        # error more than 4 elements
        haversine((1, 1, 1, 1, 1, 1))


def test_with_nparray():
    arr = np.array(
        (
            (42.3521, -72.1235, 70.612, 72.1260),
            (0.0, 0.0, 1.0, 0.0),
        )
    )
    res = haversine(arr)
    expect = [7130580.307935911, 111195.0802335329]
    assert isinstance(res, np.ndarray)
    assert pytest.approx(res[0]) == expect[0]
    assert pytest.approx(res[1]) == expect[1]


def test_slow_haversine():
    res = slow_haversine(42.3521, -72.1235, 70.612, 72.1260)
    assert pytest.approx(res) == 7130580.307935911


def test_readme_examples():
    london = (51.51, -0.12)
    paris = (48.85, 2.35)
    new_york = (40.7, -74.2)

    # calcate one distance from coordinate pairs
    result = haversine(london, paris)
    assert pytest.approx(result) == 344073.9

    # Calculate multiple distances with lists
    result = haversine([(london, paris), (new_york, london)])
    assert isinstance(result, list)
    assert pytest.approx(result[0]) == 344073.9
    assert pytest.approx(result[1]) == 5584284.1

    # or a numpy array with shape [4, n]
    result = haversine(np.array([(*london, *paris), (*new_york, *london)]))
    assert isinstance(result, np.ndarray)
    assert pytest.approx(result[0]) == 344073.9
    assert pytest.approx(result[1]) == 5584284.1


fast_haversine/slow.py
import math


def slow_haversine(lat1: float, lng1: float, lat2: float, lng2: float) -> float:
    """A pure python implementation of Haversine distance

    This function is implemented following the ArcGIS documentation at
    https://bit.ly/3wKKYo1. It is provided here as a convenience to perform
    benchmarks against the Rust implementation.

    Parameters
    ----------
    lat1 : float
        First point latitude
    lng1 : float
        First point longitude
    lat2 : float
        Second point latitude
    lng2 : float
        Second point longitude

    Returns
    -------
    float
        Haversine distance in metres
    """
    R = 6371008.8  # radius of Earth in meters
    phi_1 = math.radians(lat1)
    phi_2 = math.radians(lat2)

    delta_phi = math.radians(lat2 - lat1)
    delta_lambda = math.radians(lng2 - lng1)

    a = (
        math.sin(delta_phi / 2.0) ** 2
        + math.cos(phi_1) * math.cos(phi_2) * math.sin(delta_lambda / 2.0) ** 2
    )

    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))

    return R * c


python/fast_geodist/__init__.py
from itertools import chain
from typing import List, Tuple, Union

import numpy as np

from ._fast_geodist import haversine as _haversine
from ._fast_geodist import haversine_array

__all__ = ["haversine"]
__version__ = "0.4.0"


def haversine(
    *args: Union[
        np.ndarray,
        Tuple[Tuple[float, float], Tuple[float, float]],
        List[Tuple[Tuple[float, float], Tuple[float, float]]],
    ]
) -> Union[np.ndarray, List[float], float]:
    """Calculate the Haversine distance between pairs of latitudes and longitudes

    Parameters
    ----------
    args
        If input is a 2D numpy array with 4 columns containing lat/lon pairs, e.g.
        np.array([lat1, lon1, lat2, lon2], [...]). For each row the distance
        will be calculated.
        If the input is list a of tuples [((lat1, lon1), (lat2, long2))], the distance
        for each pair will be calculated.
        If the input is just a pair of tuples (lat1, lon1), (lat2, long2) the distance
        will be returned between the coordinates.

    Returns
    -------
    Union[np.ndarray, List[float], float]
        Haversine distance in metres, either a float or array/list of floats based on
        the input that was given.
    """
    if len(args) == 1:
        if isinstance(args[0], list):
            # must have an even number of coordinates (i.e. pairs)
            coords = args[0]
            if len(coords) % 2 != 0:
                raise ValueError(
                    f"An odd number of coordinate pairs was supplied : {len(coords)}"
                )

            # reshape and cast to a numpy array, return a list
            # TODO do this without going through numpy
            arr = np.array([list(chain(*c)) for c in coords])
            return list(haversine_array(arr))

        elif isinstance(args[0], np.ndarray):
            if args[0].ndim != 2 and args[0].shape[1] != 4:
                raise ValueError(
                    f"Input array must have shape [n, 4], got {args[0].shape}"
                )
            return haversine_array(args[0])

        else:
            raise NotImplementedError

    elif len(args) == 2:
        # should be a tuple of size 2 with 2 coordinate pairs
        if not isinstance(args, tuple):
            raise NotImplementedError(
                "Input must be a numpy array, tuple, or list of tuples"
            )

        if len(args) == 2:
            return _haversine(*args[0], *args[1])

    raise NotImplementedError(f"No implementation for {type(args)}")


src/lib.rs
use ndarray::parallel::prelude::{IndexedParallelIterator, IntoParallelIterator, ParallelIterator};
use ndarray::{Array1, Array2, ArrayView2, Axis};
use numpy::{IntoPyArray, PyArray1, PyReadonlyArray2};
use pyo3::prelude::{pyfunction, pymodule, wrap_pyfunction, PyModule, PyResult, Python};

/// Mean radius of Earth in meters
/// This is the value recommended by the IUGG:
/// Moritz, H. (2000). Geodetic Reference System 1980. Journal of Geodesy, 74(1), 128–133. doi:10.1007/s001900050278
/// "Derived Geometric Constants: mean radius" (p133)
/// https://link.springer.com/article/10.1007%2Fs001900050278
/// https://sci-hub.se/https://doi.org/10.1007/s001900050278
/// https://en.wikipedia.org/wiki/Earth_radius#Mean_radius
/// Source: https://github.com/georust/geo/blob/main/geo/src/lib.rs
const MEAN_EARTH_RADIUS: f64 = 6371008.8;

/// Calculate the Haversine (or Great Circle) distnace between two points
pub fn haversine_distance(lat1: f64, lng1: f64, lat2: f64, lng2: f64) -> f64 {
    // Following https://github.com/georust/geo/blob/main/geo/src/algorithm/haversine_distance.rs
    // Lat = Y, Lng = X
    let theta1 = lat1.to_radians();
    let theta2 = lat2.to_radians();

    let delta_theta = (lat2 - lat1).to_radians();
    let delta_lambda = (lng2 - lng1).to_radians();

    let a = (delta_theta / 2.0).sin().powi(2)
        + theta1.cos() * theta2.cos() * (delta_lambda / 2.0).sin().powi(2);

    let c = 2.0 * a.sqrt().asin();
    MEAN_EARTH_RADIUS * c
}

pub fn haversine_distance_array(x: &ArrayView2<f64>) -> Array1<f64> {
    let mut distances = Vec::with_capacity(x.shape()[0]);
    x.axis_iter(Axis(0))
        .into_par_iter()
        .map(|row| haversine_distance(row[0], row[1], row[2], row[3]))
        .collect_into_vec(&mut distances);

    Array1::from(distances)
}

pub fn haversine_combinations(arr1: &ArrayView2<f64>, arr2: &ArrayView2<f64>) -> Array2<f64> {
    // TODO panic if arr1 and arr2 aren't the same shape
    let out = Array2::zeros((arr1.shape()[0], arr2.shape()[0]));

    // TODO implement
    out
}

/// Python wrappers
#[pyfunction]
fn haversine(lat1: f64, lng1: f64, lat2: f64, lng2: f64) -> PyResult<f64> {
    Ok(haversine_distance(lat1, lng1, lat2, lng2))
}

#[pyfunction]
fn haversine_array<'py>(py: Python<'py>, x: PyReadonlyArray2<f64>) -> &'py PyArray1<f64> {
    let res = haversine_distance_array(&x.as_array());
    res.into_pyarray(py)
}

/// Python module
#[pymodule]
fn _fast_geodist(_py: Python, m: &PyModule) -> PyResult<()> {
    m.add_function(wrap_pyfunction!(haversine, m)?)?;
    m.add_function(wrap_pyfunction!(haversine_array, m)?)?;
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use approx::assert_relative_eq;
    use ndarray::array;

    #[test]
    fn test_haversine_1() {
        let res = haversine_distance(0.0, 0.0, 1.0, 0.0);
        let expect = 111195.0802335329_f64;
        assert_relative_eq!(res, expect, epsilon = 1.0e-6)
    }

    #[test]
    fn test_haversine_2() {
        let res = haversine_distance(42.3521, -72.1235, 70.612, 72.1260);
        let expect = 7130580.307935911_f64;
        assert_relative_eq!(res, expect, epsilon = 1.0e-6)
    }

    #[test]
    fn test_haversine_array() {
        let arr = array![[42.3521, -72.1235, 70.612, 72.1260], [0.0, 0.0, 1.0, 0.0]];

        let res = haversine_distance_array(&arr.view());
        let expect = [7130580.307935911, 111195.0802335329];

        assert_relative_eq!(res[0], expect[0], epsilon = 1.0e-6);
        assert_relative_eq!(res[1], expect[1], epsilon = 1.0e-6);
    }

    #[test]
    fn test_haversine_comb() {
        let arr1 = array![[45.7597, 4.8422], [51.509865, -0.118092]];
        let arr2 = array![[48.8567, 2.3508], [40.7033962, -74.2351462]];

        let res = haversine_combinations(&arr1.view(), &arr2.view());
        let expect = array![[392217.25956, 343374.55271], [6163436.38211, 5586484.47423]];

        for i in 0..expect.shape()[0] {
            for j in 0..expect.shape()[1] {
                assert_relative_eq!(res[(i, j)], expect[(i, j)], epsilon = 1.0e-6);
            }
        }
    }
}
